torch/fx/passes/infra/partitioner.py

```diff
@@ -81,6 +82,9 @@ def dfs_find_cycle(node):
                     # hit a node in a partition through DFS, we need to
                     # traverse all nodes in the partition to properly reflect
                     # dependencies after the fusion
+                    if assignment[node] in visited_partitions:
```

Comment:
This doesn't feel right.

We have `visited_partitions.add(assignment[node])` before recursively entering other nodes in the partition, so we would not further traverse to the real `outputs` of the partition group.

We did hit this issue, there's a mistake in the existing logic where we are needlessly looping inside the partition.

This might be an easier fix if we want to reduce the recursion depth: https://github.com/csarofeen/pytorch/commit/94d159f5f1306fbfff4ae43ccd196a0675317889#diff-f2f9dd3903fd99955732eb694941fea0cb7301a58d59554787f3311d417e5615R87

Reply:
We are aware of the problem here, I didn't push the fix to upstream, since python recursion limit is only 1000. If we have a large graph, we might reasonably go over that limit. So we might as well switch to an iterative approach.

I do plan on refactor this to do that, but our hands are pretty tight at this moment. I don't see us hitting a graph deeper than 1000 layers, so we'll get around and refactor that later :laughing:

Reply:
Oops, never mind. I realized that we are traversing all nodes in this partition, so we should be alright here.  

I don't even remember my own code. :facepalm:

Reply:
Yes. If a partition is already traversed once and no cycle is found, we don't traverse it again.